*: add the doc of DM argument description (#763)
pingcap/docs: tools/dm-task-config-argument-description.md

​    table-pattern: "t_*"
​    target-schema: "test"
​    target-table: "t"
```

Description: Synchronizes the upstream table data that matches `schema-pattern`/`table-pattern` to the downstream `target-schema`/`target-table`. You can set the routing rule at the schema/table level. 

Taking the above code block as an example:

- Schema level: Synchronizes all the upstream tables that match the `test_*` schema to the downstream `test` schema.  

    For example, `schema: test_1 - tables [a, b, c]`  =>  `schema:test -  tables [a, b, c]`

- Table level: Synchronizes the `t_*` matched upstream tables with `test_*` matched schema to the downstream `schema:test table:t` table.

> **Notes:** 
>
> - The `table level` rule has a higher priority than the `schema level` rule.
> - You can set one routing rule at most at one level.

## Black white list rule

```
instance:						
​    do-dbs: ["~^test.*", "do"]         # Starts with "~", indicating it is a regular expression
​    ignore-dbs: ["mysql", "ignored"]
​    do-tables:
​    - db-name: "~^test.*"
​      tbl-name: "~^t.*"
​    - db-name: "do"
​      tbl-name: "do"
​    ignore-tables:
​    - db-name: "do"
​      tbl-name: "do"
```

> Description: The black white list filter rule for the upstream database instance table.

Reviewer: ```suggestion
The black and white lists filtering rule of the upstream database instances is similar to MySQL replication-rules-db/tables.
```

The filter process is as follows:

1. Filter at the schema level:

    - If `do-dbs` is not empty, judge whether a matched schema exists in `do-dbs`. 
    
        - If yes, continue to filter at the table level.
        - If not, ignore it and exit.

    - If `do-dbs` is empty, and `ignore-dbs` is not empty, judge whether a matched schema exits in `ignore-dbs`. 
    
        - If yes, ignore it and exit.
        - If not, continue to filter at the table level.

    - If both `do-dbs` and `ignore-dbs` are empty, continue to filter at the table level.

2. Filter at the table level:
    
    1. If `do-tables` is not empty, judge whether a matched rule exists in `do-tables`.  
    
        - If yes, exit and execute the statement.
        - If not, continue to the next step.

    2. If `ignore tables` is not empty, judge whether a matched rule exists in `ignore-tables`.
    
        - If yes, ignore it and exit.
        - If not, continue to the next step.

    3. If `do-tables` is not empty, ignore it and exit. Otherwise, exit and execute the statement.

> ## Filter rules of binlog events

Reviewer: ```suggestion
## Filtering rules of binlog events
```

    | `none ddl`        |      | Includes none of the DDL events below |
    | `none dml`        |      | Includes none of the DML events below |
    | `insert`          | DML  | The `INSERT` DML event              |
    | `update`          | DML  | The `UPDATE` DML event              |
    | `delete`          | DML  | The `DELETE` DML event              |
    | `create database` | DDL  | The `CREATE DATABASE` DDL event         |
    | `drop database`   | DDL  | The `DROP DATABASE` DDL event           |
    | `create table`    | DDL  | The `CREATE TABLE` DDL event      |
    | `create index`    | DDL  | The `CREATE INDEX` DDL event          |
    | `drop table`      | DDL  | The `DROP TABLE` DDL event              |
    | `truncate table`  | DDL  | The `TRUNCATE TABLE` DDL event          |
    | `rename table`    | DDL  | The `RENAME TABLE` DDL event            |
    | `drop index`      | DDL  | The `DROP INDEX` DDL event           |
    | `alter table`     | DDL  | The `ALTER TABLE` DDL event           |

- `sql-pattern` 

    - Filters a specific DDL SQL statement. 
    - The matching rule supports using an regular expression, for example, `"^DROP\\s+PROCEDURE"`.
    

> > **Note:** If `sql-pattern` is empty, no filtering operation is performed. For the filter rules, see the `action` description.

Reviewer: ```suggestion
    > **Note:** If `sql-pattern` is empty, no filtering operation is performed. For the filter rules, see the `action` description.
```

---

> title: Task Configuration Argument Description of Data Migration

Reviewer 2: I think this title is too verbose.  How does "Data Migration Task Configuration Options" sound?

summary: This document introduces some arguments in task configuration of Data Migration.
category: tools
---

# Task Configuration Argument Description of Data Migration

> This document introduces some arguments in task configuration of Data Migration. 

Reviewer 2: "This document introduces the configuration options that apply to Data Migration tasks." ?

## `task-mode`

- String
- The task mode of data migration to be executed 
- Value: `full`/`incremental`/`all`

>     - `full`: Only makes a full backup of the upstream database and then makes a full import to the downstream database.

Reviewer 3: how about `and then restore them to the downstream database`

Author: restore it?

>     - `incremental`: Only synchronizes the incremental data of the upstream database to the downstream database using the binlog.

Reviewer 3: how about change `the incremental data` to `Binlog`

## Routing rule

```
# `schema-pattern`/`table-pattern` uses the wildcard matching rule
schema level:
​    schema-pattern: "test_*"
​    target-schema: "test"

table level:
​    schema-pattern: "test_*"
​    table-pattern: "t_*"
​    target-schema: "test"
​    target-table: "t"
```

Description: Synchronizes the upstream table data that matches `schema-pattern`/`table-pattern` to the downstream `target-schema`/`target-table`. You can set the routing rule at the schema/table level. 

Taking the above code block as an example:

- Schema level: Synchronizes all the upstream tables that match the `test_*` schema to the downstream `test` schema.  

    For example, `schema: test_1 - tables [a, b, c]`  =>  `schema:test -  tables [a, b, c]`

- Table level: Synchronizes the `t_*` matched upstream tables with `test_*` matched schema to the downstream `schema:test table:t` table.

> **Notes:** 
>
> - The `table level` rule has a higher priority than the `schema level` rule.

> > - You can set one configuration rule at most at one level.

Reviewer 3: `You can set one route rule at most at one level.`

## Black white list rule

```
instance:						
​    do-dbs: ["~^test.*", "do"]         # Starts with "~", indicating it is a regular expression
​    ignore-dbs: ["mysql", "ignored"]
​    do-tables:
​    - db-name: "~^test.*"
​      tbl-name: "~^t.*"
​    - db-name: "do"
​      tbl-name: "do"
​    ignore-tables:
​    - db-name: "do"
​      tbl-name: "do"
```

> Description: The black white list filter rule for the upstream database instance table. This rule is similar to `replication-rules-db`/`replication-rules-table` of MySQL.

Reviewer 2: MySQL replication rules for replicate-do-db are very complicated :(  I would be careful just saying it is the same: https://dev.mysql.com/doc/refman/8.0/en/replication-options-slave.html#option_mysqld_replicate-do-db

It is likely the same when in RBR, but old DBAs will be familiar with SBR which has strange semantics (different DB selected, transaction spanning dbs).

Author: @morgo I will remove "This rule is similar to `replication-rules-db`/`replication-rules-table` of MySQL."

    
        - If yes, ignore it and exit.
        - If not, continue to filter at the table level.

    - If both `do-dbs` and `ignore-dbs` are empty, continue to filter at the table level.

2. Filter at the table level:
    
    1. If `do-tables` is not empty, judge whether a matched rule exists in `do-tables`.  
    
        - If yes, exit and execute the statement.
        - If not, continue to the next step.

    2. If `ignore tables` is not empty, judge whether a matched rule exists in `ignore-tables`.
    
        - If yes, ignore it and exit.
        - If not, continue to the next step.

    3. If `do-tables` is not empty, ignore it and exit. Otherwise, exit and execute the statement.

## Filter rules of binlog events

```
# table level
user-filter-1:
​    schema-pattern: "test_*"     # `schema-pattern`/`table-pattern` uses the wildcard matching rule.
​    table-pattern: "t_*"
​    events: ["truncate table", "drop table"]
​    sql-pattern: ["^DROP\\s+PROCEDURE", "^CREATE\\s+PROCEDURE"]
​    action: Ignore

# schema level
user-filter-2:
​    schema-pattern: "test_*"
​    events: ["All DML"]
​    action: Do
```

> For the matching rules, see

Reviewer 3: remove this line?

> Description: Configures the filter rules of black white list for binlog events of the upstream tables and DDL SQL statements that match `schema-pattern`/`table-pattern`.

Reviewer 3: s/black white list/filter rule

    | `none ddl`        |      | Includes none of the DDL events below |
    | `none dml`        |      | Includes none of the DML events below |
    | `insert`          | DML  | The `INSERT` DML event              |
    | `update`          | DML  | The `UPDATE` DML event              |
    | `delete`          | DML  | The `DELETE` DML event              |
    | `create database` | DDL  | The `CREATE DATABASE` DDL event         |
    | `drop database`   | DDL  | The `DROP DATABASE` DDL event           |
    | `create table`    | DDL  | The `CREATE TABLE` DDL event      |
    | `create index`    | DDL  | The `CREATE INDEX` DDL event          |
    | `drop table`      | DDL  | The `DROP TABLE` DDL event              |
    | `truncate table`  | DDL  | The `TRUNCATE TABLE` DDL event          |
    | `rename table`    | DDL  | The `RENAME TABLE` DDL event            |
    | `drop index`      | DDL  | The `DROP INDEX` DDL event           |
    | `alter table`     | DDL  | The `ALTER TABLE` DDL event           |

- `sql-pattern` 

    - Filters a specific DDL SQL statement. 
    - The matching rule supports using an regular expression, for example, `"^DROP\\s+PROCEDURE"`.

>     - **Note:** If `sql-pattern` is empty, no filtering operation is performed. For the filter rules, see the `action` description.

Reviewer 3: `Note` should not be under `sql-pattern`

- `action` 

    - String (`Do`/`Ignore`)
    - For rules that match `schema-pattern`/`table-pattern`, judge whether the DDL statement is in the events of the rule or `sql-pattern`:
        

>         - If the DDL statement is in the events of the rule or `sql-pattern`, execute `Ignore` (black list).

Reviewer 3: if `action = Ignore`,  execute `Ignore`; otherwise execute `Do`

>         - If the DDL statement is not in the events of the rule or `sql-pattern` (not empty), execute `Do` (white list). 

Reviewer 3: if action = Ignore, execute `Do`; otherwise execute `Ignore`
